### PR TITLE
Improve charging_station.lua (closes #451)

### LINKED ISSUE #451: MTN: Charging station is effectively O(n^2)
https://github.com/ComfyFactory/ComfyFactorio/blob/cd2cda7a0e7a92d76180cd41bd2692b21e140fc4/maps/mountain_fortress_v3/charging_station.lua#L76-L81

However the n is low. It is: `O( #dischargedEquip * #nearbyAccumulators )`

To charge all equipment from all available accumulators we would need to iterate both at least once. Assuming that fetching nearby accumulator entities is a more expensive operations, I would rewrite it like this:

1. Iterate all equipment
2. Add discharged items to table, count up needed power
3. Iterate nearby accumulators. Count up available power, break return if >= needed power
4. Iterate saved discharged items and charge them based on available power

### PR DESCRIPTION
**Pull request by vadcx**
https://github.com/ComfyFactory/ComfyFactorio/issues/451

Improve algorithmic complexity as proposed. The charging button now only iterates ONCE over entire inventory and ONCE over all nearby accumulators, breaking early if enough energy was supplied.

Add support for modded accumulators. The code will check maximum energy capacity of type=accumulator entities nearby for this to work.

Add code annotations.

Improve player-facing messages.

### All Submissions:

- [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Tested Changes:

1. [ ] Have you lint your code (lua lint) locally prior to submission?

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: maps/mountain_fortress_v3/charging_station.lua [hunk 1]
```diff
@@ -34,61 +34,93 @@ local function draw_charging_gui(player, activate_custom_buttons)
...
+
+            local charge_delta = math.min(accu_energy_available, power_needs)
+
+            accu.energy = accu.energy - charge_delta
+            power_drained = power_drained + charge_delta
+            power_needs = power_needs - charge_delta
         end
     end
+
+    -- can anybody explain why the original author multiplied and divided by 1?
```
Comment (hanakocz) [inline maps/mountain_fortress_v3/charging_station.lua]: Originally there was a global coefficient so map can make charging more or less expensive, but it wasn't really used and thus got substituted to 1 while the module got moved from raw global variables into tokens

Comment (vadcx) [reply maps/mountain_fortress_v3/charging_station.lua]: Cool thanks!

**Comment by vadcx** [comment]
Silly bot, no recent activity doesn't mean the problem ceased to exist. I cannot reopen neither the issue nor the PR. I will keep pinging then.

**Review by Quadrum1 — COMMENTED** [review]
LGTM

**Comment by Quadrum1** [comment]
I agree, stalebot automatically closing tickets seems odd.
Sadly, it seems I don't have the permissions to reopen this ticket.

**Pull request closed without merging**